src/client/datascience/context/telemetry.ts

```diff
@@ -0,0 +1,139 @@
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License.
+
+import { Uri } from 'vscode';
+import { getOSType } from '../../common/utils/platform';
+import { KernelConnectionMetadata } from '../jupyter/kernels/types';
+import * as hashjs from 'hash.js';
+import { Resource } from '../../common/types';
+import { IEventNamePropertyMapping, sendTelemetryEvent, sendTelemetryWhenDone } from '../../telemetry';
+import { StopWatch } from '../../common/utils/stopWatch';
+import { ResourceSpecificTelemetryProperties } from './types';
+import { isErrorType } from '../../common/errors/errorUtils';
+import { CancellationError } from '../../common/cancellation';
+import { TimedOutError } from '../../common/utils/async';
+import { JupyterInvalidKernelError } from '../jupyter/jupyterInvalidKernelError';
+import { JupyterWaitForIdleError } from '../jupyter/jupyterWaitForIdleError';
+import { JupyterKernelPromiseFailedError } from '../jupyter/kernels/jupyterKernelPromiseFailedError';
+import { IpyKernelNotInstalledError, KernelDiedError } from '../kernel-launcher/types';
+import { JupyterSessionStartError } from '../baseJupyterSession';
+import { JupyterConnectError } from '../jupyter/jupyterConnectError';
+import { JupyterInstallError } from '../jupyter/jupyterInstallError';
+import { JupyterSelfCertsError } from '../jupyter/jupyterSelfCertsError';
+
+type ContextualTelemetryProps = {
+    kernelConnection: KernelConnectionMetadata;
+    wasJupyterAutoStarted: boolean;
+    kernelDied: boolean;
+};
+
+const trackedInfo = new Map<string, ResourceSpecificTelemetryProperties>();
+const currentOSType = getOSType();
+
+export function getKernelFailureReason(error: Error) {
```

Comment:
What about parsing the error returned from the kernel (the python exception)? Maybe you did that elsewhere.

Reply:
Didn't find it elsewhere. That was one of the properties I thought we were going to try sending. Like if it was pyzmq, or pywin32 or DLL load or zmqconstants not found, etc.

Reply:
Oh maybe you haven't gotten to it yet.

Reply:
Yes, still more to be done

Reply:
Lots more.

Reply:
> Didn't find it elsewhere. That was one of the properties I thought we were going to try sending. Like if it was pyzmq, or pywin32 or DLL load or zmqconstants not found, etc.

 
@rchiodo looking at the past issues, users managed to re-install pyzmq or ipykernel to resolve these. Hence I believe just getting the versions of these related packages might be sufficient & the type of python they are using (conda, etc).
E.g. when using conda and it isn't activated, we get DLL load errors.

Reply:
I.e. i'm suggesting we not check the stderr for now.
& add that later if required (if we still find many users failing there & the version information isnt' sufficient).

Reply:
I believe we need the stderr to even tell if people are hitting these common errors.  Version information isn't sufficient for these things (for example, the pyzmq problem. Requires uninstall and reinstall - same version on both, something else messed it up) and there may be other ones we don't even know about that are common. 

What's the downside to adding the stderr?

Reply:
Stderr can contain PII was the consideration. It's not a structured error or exception, just raw output. Could contain stuff like paths or folder names, env vars, other stuff.

Reply:
I think we can workaround that by just sending the filename and line number of the topmost frame. That should be non PII. We could even double check if it was a site packages thing or not.

Reply:
filename and linenumber should be unique enough to tell what error it is.

Reply:
Sure, will work on the stderr serpately then.
Challenge will be trying to tie it to the same telemetry, will be challenging, worst case we can have a seprate telemetry event for the stderr contents